ADAL/src/workplacejoin/mac/ADClientCertAuthHandler.m

```diff
@@ -85,4 +90,152 @@ + (BOOL)handleChallenge:(NSURLAuthenticationChallenge *)challenge
+                                   host:(NSString *)host
+                          correlationId:(NSUUID *)correlationId
+{
+    NSDictionary *query =
+    @{
+      (id)kSecClass : (id)kSecClassIdentity,
+      (id)kSecMatchIssuers : issuers,
+      (id)kSecMatchLimit : (id)kSecMatchLimitAll,
+      };
+    
+    CFTypeRef result = NULL;
+    
+    OSStatus status = SecItemCopyMatching((CFDictionaryRef)query, &result);
+    if (status == errSecItemNotFound)
+    {
+        AD_LOG_INFO(@"No certificate found matching challenge", correlationId, nil);
+        return nil;
+    }
+    else if (status != errSecSuccess)
+    {
+        AD_LOG_ERROR(([NSString stringWithFormat:@"Failed to find identity matching issuers with %d error.", status]), status, correlationId, nil);
+        return nil;
+    }
+    
+    return [ADCertificateChooserHelper showCertSelectionSheet:(__bridge NSArray *)result host:host correlationId:correlationId];
+}
+
+
++ (BOOL)handleChallenge:(NSURLAuthenticationChallenge *)challenge
+                session:(NSURLSession *)session
+                   task:(NSURLSessionTask *)task
+               protocol:(ADURLProtocol *)protocol
+      completionHandler:(ChallengeCompletionHandler)completionHandler;
+{
+#pragma unused(session)
+#pragma unused(task)
+    
+    NSUUID *correlationId = protocol.context.correlationId;
+    NSString *host = challenge.protectionSpace.host;
+    AD_LOG_INFO_F(@"Attempting to handle client certificate challenge", correlationId, @"host: %@", host);
+    
+    // See if this is a challenge for the WPJ cert.
+    NSArray<NSData*> *distinguishedNames = challenge.protectionSpace.distinguishedNames;
+    if ([self isWPJChallenge:distinguishedNames])
+    {
+        return [self handleWPJChallenge:challenge protocol:protocol completionHandler:completionHandler];
+    }
+    
+    // Otherwise check if a preferred identity is set for this host
+    SecIdentityRef identity = SecIdentityCopyPreferred((CFStringRef)host, NULL, (CFArrayRef)distinguishedNames);
+    if (identity != NULL)
+    {
+        AD_LOG_INFO(@"Using preferred identity", correlationId, nil);
+    }
+    else
+    {
+        // If not prompt the user to select an identity
+        identity = [self promptUserForIdentity:distinguishedNames host:host correlationId:correlationId];
+        if (identity == NULL)
+        {
+            // If no identity comes back then we can't handle the request
+            return NO;
+        }
+        
+        // Adding a retain count to match the retain count from SecIdentityCopyPreferred
+        CFRetain(identity);
+        AD_LOG_INFO(@"Using user selected certificate", correlationId, nil);
+    }
+    
+    SecCertificateRef cert = NULL;
+    OSStatus status = SecIdentityCopyCertificate(identity, &cert);
+    if (status != errSecSuccess)
+    {
+        CFRelease(identity);
+        AD_LOG_ERROR(@"Failed to copy certificate from identity", AD_ERROR_UNEXPECTED, correlationId, nil);
+        return NO;
```

Comment:
Just curious, if it's possible that there will be multiple authentication options available and one of them is cert (e.g if we cannot handle cert, fallback to NTLM)?